wdl-lint/src/rules/shellcheck.rs

```diff
@@ -0,0 +1,424 @@
+    match output.status.code() {
+        Some(0) | Some(1) => serde_json::from_slice::<Vec<ShellCheckDiagnostic>>(&output.stdout)
+            .context("deserializing STDOUT from `shellcheck` process"),
+        Some(code) => bail!("unexpected `shellcheck` exit code: {}", code),
+        None => bail!("the `shellcheck` process appears to have been interrupted"),
+    }
+}
+
+/// Runs ShellCheck on a command section and reports diagnostics.
+#[derive(Default, Debug, Clone, Copy)]
+pub struct ShellCheckRule;
+
+impl Rule for ShellCheckRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that command blocks are free of ShellCheck violations."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "ShellCheck (https://shellcheck.net) is a static analysis tool and linter for sh / bash. \
+         The lints provided by ShellCheck help prevent common errors and pitfalls in your scripts. \
+         Following its recommendations will increase the robustness of your command sections."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Correctness, Tag::Portability])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::CommandSectionNode,
+        ])
+    }
+}
+
+/// Convert a WDL `Placeholder` to a bash variable declaration.
+///
+/// Returns "WDL" + <placeholder length - 6> random alphnumeric characters.
+/// The returned value is shorter than the placeholder by 3 characters so
+/// that the caller may pad with other characters as necessary
+/// depending on whether or not the variable needs to be treated as a
+/// declaration, expansion, or literal.
+fn to_bash_var(placeholder: &Placeholder) -> String {
+    let placeholder_len: usize = placeholder.syntax().text_range().len().into();
+    // don't start variable with numbers
+    let mut bash_var = String::from("WDL");
+    bash_var.push_str(
+        &Alphanumeric.sample_string(&mut rand::thread_rng(), placeholder_len.saturating_sub(6)),
+    );
+    bash_var
+}
+
+/// Retrieve all input and private declarations for a task.
+fn gather_task_declarations(task: &TaskDefinition) -> HashSet<String> {
+    let mut decls = HashSet::new();
+    if let Some(input) = task.input() {
+        for decl in input.declarations() {
+            decls.insert(decl.name().as_str().to_owned());
+        }
+    }
+
+    for decl in task.declarations() {
+        decls.insert(decl.name().as_str().to_owned());
+    }
+    decls
+}
+
+/// Creates a "ShellCheck lint" diagnostic from a `ShellCheckDiagnostic`
+fn shellcheck_lint(diagnostic: &ShellCheckDiagnostic, span: Span) -> Diagnostic {
+    let label = format!(
+        "SC{}[{}]: {}",
+        diagnostic.code, diagnostic.level, diagnostic.message
+    );
+    Diagnostic::note(&diagnostic.message)
+        .with_rule(ID)
+        .with_label(label, span)
```

Comment:
This repeats the `diagnostic.message` twice. Is that desirable? Should we omit it from the `label` as being redundant, or do we like it appearing twice?

Reply:
I'm conflicted. Pulling an example from below, I think the initial instance by the `note` is fine. The second provides the additional info link, which could be helpful. I do notice that the span isn't highlighted correctly. I also think the `fix` needs to be something else.
```
note[ShellCheck]: You need a space before the ].
   ┌─ tests/lints/shellcheck-error/source.wdl:18:25
   │
18 │       if [ -f "$broken"]
   │                         ^
   │                         
   │                         SC1020[error]: You need a space before the ].
   │                         more info: https://www.shellcheck.net/wiki/SC1020
   │
   = fix: address the diagnostic as recommended in the message
```

Reply:
Re: the message

@peterhuene  and I discussed this some in the Zulip. This initially came about because Gauntlet requires all diagnostic messages to be unique. The solution I came up with was to include the SC###[<severity>] in the message, but only the message in the label. But happy to rework. Agree the fix could be much more useful.

Reply:
So it's the underlying `shellcheck` that returns the column position after `]`. That's good to know. It still seems like the highlight should be one position earlier, so that it flags the `]`.

Reply:
This thread got a little sidetracked. Bringing up the original Q again, are we ok with the SC message being repeated? Maybe I'm lacking in imagination here, but I don't really see a way to get rid of the duplication that still looks ok.

I like it at the top of the diagnostic and that ensures uniqueness for gauntlet. Removing it from the label leaves that part lacking... I'm stumped.
I'm ok with it appearing twice, but that's mostly because I can't think of an alternative I like. Open to any ideas here.

Reply:
I'm fine with the duplication. We can always revisit it in the future.